master/internal/stream/projects.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// SeqNum gets the SeqNum from a ProjectMsg.
+func (pm *ProjectMsg) SeqNum() int64 {
+	return pm.Seq
+}
+
+// UpsertMsg creates a Project stream upsert message.
+func (pm *ProjectMsg) UpsertMsg() stream.UpsertMsg {
+	return stream.UpsertMsg{
+		JSONKey: ProjectsUpsertKey,
+		Msg:     pm,
+	}
+}
+
+// DeleteMsg creates a Project stream delete message.
+func (pm *ProjectMsg) DeleteMsg() stream.DeleteMsg {
+	deleted := strconv.FormatInt(int64(pm.ID), 10)
+	return stream.DeleteMsg{
+		Key:     ProjectsDeleteKey,
+		Deleted: deleted,
+	}
+}
+
+// ProjectSubscriptionSpec is what a user submits to define a project subscription.
+type ProjectSubscriptionSpec struct {
+	WorkspaceIDs []int `json:"workspace_ids"`
+	ProjectIDs   []int `json:"project_ids"`
+	Since        int64 `json:"since"`
+}
+
+// ProjectCollectStartupMsgs collects ProjectMsg's that were missed prior to startup.
+func ProjectCollectStartupMsgs(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec ProjectSubscriptionSpec,
+) (
+	[]stream.PreparableMessage, error,
+) {
+	var out []stream.PreparableMessage
+
+	if len(spec.WorkspaceIDs) == 0 && len(spec.ProjectIDs) == 0 {
+		// empty subscription: everything known should be returned as deleted
+		out = append(out, stream.DeleteMsg{
+			Key:     ProjectsDeleteKey,
+			Deleted: known,
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetProjectStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	var accessScopes []model.AccessScopeID
+	for id, isPermitted := range accessMap {
+		if isPermitted {
+			accessScopes = append(accessScopes, id)
+		}
+	}
+
+	// step 1: calculate all ids matching this subscription
+	var data []*ProjectMsg
+	q := db.Bun().NewSelect().Model(&data).Order("id ASC")
+	q = permFilter(q, accessMap, accessScopes)
+
+	// Ignore tmf.Since, because we want appearances, which might not be have seq > spec.Since.
+	ws := stream.WhereSince{Since: 0}
+	if len(spec.WorkspaceIDs) > 0 {
+		ws.Include("workspace_id in (?)", bun.In(spec.WorkspaceIDs))
+	}
+	if len(spec.ProjectIDs) > 0 {
+		ws.Include("id in (?)", bun.In(spec.ProjectIDs))
+	}
+	q = ws.Apply(q)
+
+	err = q.Scan(ctx, &data)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	var exist []int64
+	for _, pm := range data {
+		exist = append(exist, int64(pm.ID))
+	}
+
+	// step 2: figure out what was missing and what has appeared
+	missing, appeared, err := stream.ProcessKnown(known, exist)
+	if err != nil {
+		return nil, err
+	}
+
+	appearedMsgs := make([]*ProjectMsg, 0, len(appeared))
+	if len(appeared) > 0 {
+		sort.SliceStable(appeared, func(i, j int) bool {
+			return appeared[i] < appeared[j]
+		})
+		x := 0 // tracks projectMsgs
+		y := 0 // tracks appeared IDs
+
+		for x < len(appeared) {
+			if y >= len(data) {
+				log.Errorf("appeared checkpoints are not in scanned dataset")
```

Comment:
should this be projects?
```suggestion
				log.Errorf("appeared projects are not in scanned dataset")
```